### PR TITLE
chore: Attempt to enable disallowed-methods clippy lint

### PR DESCRIPTION
**Pull request by blt**
This commit enables two new lints in the project: disallowed-methods
and missing-const-for-fn. The first is new to Rust and allows us to
reject some methods from the project if we specify the path fully in
clippy.toml. The second suggests functions that might reasonably be
const, though it has some misfires as the small handful of `allow`
introductions atests to.

disallowed-methods isn't as smooth as I'd like -- namely, I can't get
it to ding introductions of lazy_static -- but it's a start. The const
lint will help vector play better with the optimizer.

Signed-off-by: Brian L. Troutwine <brian@troutwine.us>

### CODE REVIEW

File: src/api/schema/metrics/filter.rs [hunk 1]
```diff
@@ -6,13 +6,13 @@ use crate::{
...
-use std::{collections::BTreeMap, sync::Arc};
+use std::collections::BTreeMap;
 use tokio::time::Duration;
 use tokio_stream::{Stream, StreamExt};
 
 lazy_static! {
-    static ref GLOBAL_CONTROLLER: Arc<&'static Controller> =
-        Arc::new(get_controller().expect("Metrics system not initialized. Please report."));
+    static ref GLOBAL_CONTROLLER: &'static Controller =
+        get_controller().expect("Metrics system not initialized. Please report.");
```
Comment (bruceg) [inline src/api/schema/metrics/filter.rs:15]: Oh, hey, that's a handy removal.

Comment (blt) [reply src/api/schema/metrics/filter.rs:15]: I thought so! One less indirection for sure.

File: src/sinks/util/adaptive_concurrency/tests.rs [hunk 2]
```diff
@@ -82,16 +82,17 @@ impl LimitParams {
     }
 
     fn scale(&self, level: usize) -> f64 {
-        (level - 1) as f64 * self.scale
-            + self
-                .knee_start
+        ((level - 1) as f64).mul_add(
+            self.scale,
+            self.knee_start
```
Comment (bruceg) [inline src/sinks/util/adaptive_concurrency/tests.rs:87]: Really? `mul_add` is less cognitively difficult than standard infix operators?

File: src/stats.rs [hunk 3]
```diff
@@ -6,20 +6,20 @@ pub struct Ewma {
...
+    pub const fn average(&self) -> Option<f64> {
         self.average
     }
 
     /// Update the current average and return it for convenience
     pub fn update(&mut self, point: f64) -> f64 {
         let average = match self.average {
             None => point,
-            Some(avg) => point * self.alpha + avg * (1.0 - self.alpha),
+            Some(avg) => point.mul_add(self.alpha, avg * (1.0 - self.alpha)),
```
Comment (jszwedko) [inline src/stats.rs:22]: `mul_add` is nifty

**Review by jszwedko — APPROVED** [review]
Nice, makes sense to me. I'm curious where the additional clippy lints came from though since I see you only added two additional `deny`s, were you running nightly clippy?

**Review by spencergilbert — APPROVED** [review]
Will there be a _noticeable_ difference with all the `const`? I guess I'm not entirely clear what `const` is improving specifically.

**Review by bruceg — COMMENTED** [review]
I would have liked to see the `const` in particular, and possibly the `pub(crate)` ones, separated out to see before/after comparison of release builds.

File: src/sinks/util/adaptive_concurrency/tests.rs (hunk 2 above)
Comment (jszwedko) [reply src/sinks/util/adaptive_concurrency/tests.rs:87]: I think it's more that it rounds less:

```
Fused multiply-add. Computes (self * a) + b with only one rounding error, yielding a more accurate result than an unfused multiply-add.
```

File: lib/shared/src/conversion.rs [hunk 4]
```diff
@@ -32,18 +32,18 @@ pub enum Conversion {
 #[derive(Debug, Eq, PartialEq, Snafu)]
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply lib/shared/src/conversion.rs:32]: The enum already has `Error` in the name so having `Error` in all of the variant names is redundant

https://rust-lang.github.io/rust-clippy/master/#enum_variant_names

[the remark this answers is not in the document]
Comment (spencergilbert) [reply lib/shared/src/conversion.rs:32]: 👍

**Comment by blt** [comment]
> Nice, makes sense to me. I'm curious where the additional clippy lints came from though since I see you only added two additional `deny`s, were you running nightly clippy?

I did a clippy fix with the full nursery suite enabled.

File: lib/shared/src/conversion.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (blt) [reply lib/shared/src/conversion.rs:32]: Clippy dings enums where the name of the enum appears as a suffix to the variants. So, it recommends `Error::FooError` be renamed to `Error::Foo`.

File: src/internal_events/splunk_hec.rs [hunk 5]
```diff
@@ -3,10 +3,10 @@ use metrics::counter;
 use serde_json::Error;
 
 #[cfg(feature = "sources-splunk_hec")]
-pub(crate) use self::source::*;
+pub use self::source::*;
```
[the remark this answers is not in the document]
Comment (blt) [reply src/internal_events/splunk_hec.rs:6]: Almost surely `redundant_pub_crate`. Came in as a part of clippy fix.

File: src/internal_events/aws_kinesis_firehose.rs [hunk 6]
```diff
@@ -3,7 +3,7 @@ use crate::sources::aws_kinesis_firehose::Compression;
 use metrics::counter;
 
 #[derive(Debug)]
-pub(crate) struct AwsKinesisFirehoseEventReceived {
+pub struct AwsKinesisFirehoseEventReceived {
```
[the remark this answers is not in the document]
Comment (blt) [reply src/internal_events/aws_kinesis_firehose.rs:6]: It's a good question. I'm not certain about this one specifically but we had a fair number of dings for `redundant_pub_crate` that got swept up by clippy fix.

**Comment by jszwedko** [comment]
> > Nice, makes sense to me. I'm curious where the additional clippy lints came from though since I see you only added two additional `deny`s, were you running nightly clippy?
> 
> I did a clippy fix with the full nursery suite enabled.

Ah, gotcha. I take it we don't want to start doing that in CI?

File: src/sinks/util/adaptive_concurrency/tests.rs (hunk 2 above)
Comment (blt) [reply src/sinks/util/adaptive_concurrency/tests.rs:87]: Seems about the same to me. It's something that came in with clippy fix.

**Comment by blt** [comment]
> > > Nice, makes sense to me. I'm curious where the additional clippy lints came from though since I see you only added two additional `deny`s, were you running nightly clippy?
> > 
> > 
> > I did a clippy fix with the full nursery suite enabled.
> 
> Ah, gotcha. I take it we don't want to start doing that in CI?

There'd be false positives that need manual intervention, so probably not? The big motivation here was to get the project nursery clean enough for disallow-method, the bulk of which were the clippy fixups of const changes. Sadly it doesn't seem like that lint works for macros or crates generally, so `lazy_static` lives to see another day.

**Comment by blt** [comment]
> Will there be a _noticeable_ difference with all the `const`? I guess I'm not entirely clear what `const` is improving specifically.

Not unless we were tremendously lucky and one of these call sites was a bottleneck. These are, by the look of them, low hanging fruit for clippy to call out. As const programming spreads in Rust we'll have more opportunity to build stuff at compile time, so making const what can be const is the first part in a long process of language evolution.

**Comment by blt** [comment]
> I would have liked to see the `const` in particular, and possibly the `pub(crate)` ones, separated out to see before/after comparison of release builds.

There's very few of those `pub(crate)` changes that got tossed in. I'd be happy to back them out by hand if you like. My main goal here is to have disallowed-method lint available; the const switchover and other clippy fix changes are a happy side benefit, I figured.

File: src/internal_events/aws_kinesis_firehose.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply src/internal_events/aws_kinesis_firehose.rs:6]: If they are redundant (due to the parent module being `pub(crate)` or some such) then this is good. I'm not seeing the redundancy, though. These symbols are imported with `pub use aws_kinesis_firehose::*;`, and `internal_events` is just `pub` too.

**Pull request merged**